conda_forge_tick/update_sources.py

```diff
@@ -0,0 +1,450 @@
+import abc
+import collections.abc
+import subprocess
+import re
+import logging
+import typing
+from typing import (
+    Any,
+    Optional,
+    Iterable,
+    Set,
+    Iterator,
+    List,
+)
+import yaml
+import feedparser
+import requests
+from conda.models.version import VersionOrder
+from conda_forge_tick.utils import parse_meta_yaml
+from hashing import hash_url
+
+# TODO: parse_version has bad type annotations
+from pkg_resources import parse_version
+
+CRAN_INDEX: Optional[dict] = None
+
+logger = logging.getLogger("conda-forge-tick._update_version.update_sources")
+
+def urls_from_meta(meta_yaml: "MetaYamlTypedDict") -> Set[str]:
+    source: "SourceTypedDict" = meta_yaml["source"]
+    sources: typing.List["SourceTypedDict"]
+    if isinstance(source, collections.abc.Mapping):
+        sources = [source]
+    else:
+        sources = typing.cast("typing.List[SourceTypedDict]", source)
+    urls = set()
+    for s in sources:
+        if "url" in s:
+            # if it is a list for instance
+            if not isinstance(s["url"], str):
+                urls.update(s["url"])
+            else:
+                urls.add(s["url"])
+    return urls
+
+
+def _split_alpha_num(ver: str) -> List[str]:
+    for i, c in enumerate(ver):
+        if c.isalpha():
+            return [ver[0:i], ver[i:]]
+    return [ver]
+
+
+def next_version(ver: str) -> Iterator[str]:
+    ver_split = []
+    ver_dot_split = ver.split(".")
+    n_dot = len(ver_dot_split)
+    for idot, sdot in enumerate(ver_dot_split):
+
+        ver_under_split = sdot.split("_")
+        n_under = len(ver_under_split)
+        for iunder, sunder in enumerate(ver_under_split):
+
+            ver_dash_split = sunder.split("-")
+            n_dash = len(ver_dash_split)
+            for idash, sdash in enumerate(ver_dash_split):
+
+                for el in _split_alpha_num(sdash):
```

Comment:
the cyclomatic complexity here is a bit high, isn't it?
sorry, but can you just give me a brief explanation what this function ``next_version`` is doing? to be sure that what I think it is

Reply:
I see, that is fine then, but I think we should add a ``TODO`` here
because we can parse it using the ``packaging`` or using ``distutils``
and it would be nice to add tests for it @CJ-Wright

Reply:
This function has tests: https://github.com/regro/cf-scripts/blob/master/tests/test_upstream_versions.py

We are not purely parsing a version here, plus not all versions we parse are semvers.

Reply:
See the code here: https://github.com/regro/cf-scripts/blob/master/tests/test_upstream_versions.py#L170

Reply:
Thanks for the reply Matthew, I was wondering if it had a test, good to know this .

Reply:
Very nice that we have tests for it! :)
but I think we can try to simplify it, let me try some code